runtime/lua/vim/lsp.lua

```diff
@@ -1535,5 +1535,35 @@ end
 -- Define the LspDiagnostics signs if they're not defined already.
 require('vim.lsp.diagnostic')._define_default_signs_and_highlights()
 
+
+--- Registry for client side code action commands.
+---
+--- This is an extension point for plugins to handle custom commands which are
+--- not part of the core language server protocol specification.
+---
+--- The codeAction response contains commands which are either executed with a
+--- function from this registry, or via `vim.lsp.buf.execute_command` if there is
+--- no entry for a given command.
+---
+--- This is a table where the key must be a command name and the value must be
+--- function with two arguments.
```

Comment:
hmm, how does this address the use-case from #12326 :

> Currently a plugin could choose to override the `textDocument/codeAction` callback or provide a custom `code_action` function. But then each plugin that wants to handle some command in a custom way would end-up re-implementing the code-action boilerplate and would be incompatible with each other.

how will multiple plugins add hooks to the same code action?

Reply:
Both plugins would register their command handlers within `vim.lsp.commands`.
The command keys would still be unique. So you cannot have two plugins provide the same command. But plugins for different language server wouldn't conflict.

E.g nvim-jdtls would add entries like this:

```lua
vim.lsp.commands['java.apply.workspaceEdit'] = ...
vim.lsp.commands['java.action.organizeImports'] = ...
```

and rust-tools.nvim could add rust-analyzer specific entries:


```lua
vim.lsp.commands['rust-analyzer.runSingle'] = ...
```